// react-hooks-use-modal/dist/future/index.mjs
// src/future/index.tsx
import React3, {
  useCallback,
  useEffect,
  useRef,
  useSyncExternalStore
} from "react";

// src/future/components/DefaultComponents.tsx
import React from "react";
var ModalContentDefault = ({
  title,
  description,
  children,
  contentProps
}) => {
  return /* @__PURE__ */ React.createElement("div", { ...contentProps }, title && /* @__PURE__ */ React.createElement("h2", null, title), description && /* @__PURE__ */ React.createElement("p", null, description), children);
};

// src/future/hooks/useModalConfig.ts
import { createContext, useContext } from "react";
var ModalConfigContext = createContext({});
var useModalConfig = () => {
  return useContext(ModalConfigContext);
};

// esbuild-css-modules-plugin-namespace:/tmp/tmp-1777-37f07X242LJe/react-hooks-use-modal/src/future/index.module.css.js
var digest = "268ef182bcac292131d11fb827aff8188d17f2cc1b9cf5838c83cd06935c8aee";
var css = `._dialog_jfhr7_1 {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  display: flex;
  align-items: center;
  justify-content: center;

  /* reset */
  width: 100%;
  height: 100%;
  max-width: none;
  max-height: none;
  margin: 0;
  padding: 0;
  border: none;
  background: transparent;
  inset-block-start: 0;
  inset-block-end: 0;
  inset-inline-start: 0;
  inset-inline-end: 0;
}

._dialog_jfhr7_1[open] {
  animation: _scale_jfhr7_1 0.2s cubic-bezier(0.5, 1, 0.89, 1) forwards;
}

._dialog_jfhr7_1:not([open]) {
  display: flex;
  opacity: 0;
  visibility: hidden;
  transform: scale(0.9);
  transition: opacity 0.15s cubic-bezier(0.5, 1, 0.89, 1),
    visibility 0.15s cubic-bezier(0.5, 1, 0.89, 1),
    transform 0.15s cubic-bezier(0.5, 1, 0.89, 1);
}

._dialog_jfhr7_1::backdrop {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0.5);
  animation: _fade_jfhr7_1 0.15s cubic-bezier(0.5, 1, 0.89, 1) forwards;
}

@keyframes _scale_jfhr7_1 {
  from {
    opacity: 0;
    transform: scale(0.95);
  }
  to {
    opacity: 1;
    transform: scale(1);
  }
}

@keyframes _fade_jfhr7_1 {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}
`;
(function() {
  if (typeof document === "undefined") {
    return;
  }
  if (!document.getElementById(digest)) {
    var el = document.createElement("style");
    el.id = digest;
    el.textContent = css;
    document.head.appendChild(el);
  }
})();
var index_module_css_default = { "dialog": "_dialog_jfhr7_1", "scale": "_scale_jfhr7_1", "fade": "_fade_jfhr7_1" };

// src/future/components/ModalProvider.tsx
import React2 from "react";
var ModalProvider = ({
  children,
  ...props
}) => {
  return /* @__PURE__ */ React2.createElement(ModalConfigContext.Provider, { value: props }, children);
};

// src/future/index.tsx
var useModal = (options) => {
  const modalConfig = useModalConfig();
  const initialValue = options?.initialValue ?? modalConfig.initialValue ?? false;
  const closeOnOverlayClick = options?.closeOnOverlayClick ?? modalConfig.closeOnOverlayClick ?? true;
  const ModalContent = options?.components?.ModalContent ?? modalConfig.components?.ModalContent ?? ModalContentDefault;
  const ref = useRef(null);
  const subscribe = useCallback((callback) => {
    const observer = new MutationObserver((mutations) => {
      mutations.forEach((mutation) => {
        if (mutation.attributeName === "open") {
          callback();
        }
      });
    });
    ref.current && observer.observe(ref.current, {
      attributes: true
    });
    return () => {
      observer.disconnect();
    };
  }, []);
  const isOpen = useSyncExternalStore(
    subscribe,
    () => {
      return ref.current?.open ?? false;
    },
    () => false
  );
  const open = useCallback(
    (options2 = {}) => {
      options2.isModal = options2.isModal ?? true;
      if (options2.isModal) {
        ref.current?.showModal();
      } else {
        ref.current?.show();
      }
    },
    []
  );
  const close = useCallback(() => {
    ref.current?.close();
  }, []);
  const handleOverlayClick = useCallback(
    (e) => {
      if (closeOnOverlayClick && e.target === e.currentTarget) {
        close();
      }
    },
    [closeOnOverlayClick, close]
  );
  const justMounted = useRef(true);
  useEffect(() => {
    if (justMounted.current) {
      if (initialValue === true) {
        open();
      }
      justMounted.current = false;
    }
  }, [initialValue, open]);
  const ModalWrapper = useCallback(
    ({ dialogProps, ...props }) => {
      return /* @__PURE__ */ React3.createElement(
        "dialog",
        {
          ref,
          className: [index_module_css_default.dialog, dialogProps?.className].filter(Boolean).join(" "),
          onClick: (e) => {
            dialogProps?.onClick?.(e);
            handleOverlayClick(e);
          },
          ...dialogProps
        },
        /* @__PURE__ */ React3.createElement(ModalContent, { ...props })
      );
    },
    [handleOverlayClick, ModalContent]
  );
  return [ModalWrapper, open, close, isOpen];
};
export {
  ModalProvider,
  useModal
};
